### ISSUE TITLE
Failure to create execution plan for complex query when statistics are present

### ISSUE THREAD

**Issue by hammerhead** [crate/crate#17016]
### CrateDB version

5.9.3

### CrateDB setup information

Single node

### Problem description

An `SQLParseException` is thrown for a certain complex query. The query runs successfully without statics, but fails if tables are populated and `ANALYZE` was run.

### Steps to Reproduce

```sql
CREATE TABLE IF NOT EXISTS "doc"."t1" (
   "relation_id" BIGINT,
   "parent_relation_id" BIGINT
);

CREATE TABLE IF NOT EXISTS "doc"."t2" (
   "relation_id" BIGINT
);
```

```bash
cr8 insert-fake-data -n 10 --table t1
cr8 insert-fake-data -n 50000 --table t2
```

```sql
ANALYZE;

-- reduced/anonymized query from a real-world scenario
WITH cte_1 AS (
    SELECT g.relation_id
    FROM t1 g
    WHERE g.parent_relation_id = 1
)
SELECT (
        SELECT true
        FROM t1 s
        WHERE c.relation_id = s.relation_id
       )
FROM cte_1 c
JOIN t2 p ON c.relation_id = p.relation_id;
````

### Actual Result

```
io.crate.exceptions.SQLParseException: Couldn't create execution plan from logical plan because of: Couldn't find (SELECT true FROM (s)) in SourceSymbols{inputs={relation_id=INPUT(1), relation_id=INPUT(0)}, nonDeterministicFunctions={}}:
Eval[(SELECT true FROM (s))] (rows=5000)
  └ CorrelatedJoin[(SELECT true FROM (s)), relation_id, (SELECT true FROM (s))] (rows=5000)
    └ Eval[(SELECT true FROM (s)), relation_id] (rows=5000)
      └ HashJoin[(relation_id = relation_id)] (rows=5000)
        ├ Rename[relation_id] AS p (rows=50000)
        │  └ Collect[doc.t2 | [relation_id] | true] (rows=50000)
        └ Rename[relation_id] AS c (rows=1)
          └ Rename[relation_id] AS cte_1 (rows=1)
            └ Rename[relation_id] AS g (rows=1)
              └ Collect[doc.t1 | [relation_id] | (parent_relation_id = 1::bigint)] (rows=1)
    └ SubPlan
      └ Eval[true] (rows=1)
        └ Rename[true] AS s (rows=1)
          └ Limit[2::bigint;0::bigint] (rows=1)
            └ Collect[doc.t1 | [true] | (relation_id = relation_id)] (rows=1)
	at io.crate.exceptions.SQLExceptions.esToCrateException(SQLExceptions.java:211)
	at io.crate.exceptions.SQLExceptions.prepareForClientTransmission(SQLExceptions.java:200)
	at io.crate.rest.action.SqlHttpHandler.sendResponse(SqlHttpHandler.java:172)
	at io.crate.rest.action.SqlHttpHandler.lambda$channelRead0$0(SqlHttpHandler.java:127)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2357)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:125)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:86)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.crate.protocols.http.HttpBlobHandler.channelRead0(HttpBlobHandler.java:159)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.crate.auth.HttpAuthUpstreamHandler.handleHttpChunk(HttpAuthUpstreamHandler.java:155)
	at io.crate.auth.HttpAuthUpstreamHandler.channelRead0(HttpAuthUpstreamHandler.java:92)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:289)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at org.elasticsearch.transport.netty4.Netty4InboundStatsHandler.channelRead(Netty4InboundStatsHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.lang.IllegalArgumentException: Couldn't create execution plan from logical plan because of: Couldn't find (SELECT true FROM (s)) in SourceSymbols{inputs={relation_id=INPUT(1), relation_id=INPUT(0)}, nonDeterministicFunctions={}}:
Eval[(SELECT true FROM (s))] (rows=5000)
  └ CorrelatedJoin[(SELECT true FROM (s)), relation_id, (SELECT true FROM (s))] (rows=5000)
    └ Eval[(SELECT true FROM (s)), relation_id] (rows=5000)
      └ HashJoin[(relation_id = relation_id)] (rows=5000)
        ├ Rename[relation_id] AS p (rows=50000)
        │  └ Collect[doc.t2 | [relation_id] | true] (rows=50000)
        └ Rename[relation_id] AS c (rows=1)
          └ Rename[relation_id] AS cte_1 (rows=1)
            └ Rename[relation_id] AS g (rows=1)
              └ Collect[doc.t1 | [relation_id] | (parent_relation_id = 1::bigint)] (rows=1)
    └ SubPlan
      └ Eval[true] (rows=1)
        └ Rename[true] AS s (rows=1)
          └ Limit[2::bigint;0::bigint] (rows=1)
            └ Collect[doc.t1 | [true] | (relation_id = relation_id)] (rows=1)
	at io.crate.execution.dsl.projection.builder.InputColumns.visitSelectSymbol(InputColumns.java:350)
	at io.crate.execution.dsl.projection.builder.InputColumns.visitSelectSymbol(InputColumns.java:61)
	at io.crate.expression.symbol.SelectSymbol.accept(SelectSymbol.java:94)
	at io.crate.execution.dsl.projection.builder.InputColumns.create(InputColumns.java:161)
	at io.crate.planner.operators.Eval.addEvalProjection(Eval.java:150)
	at io.crate.planner.operators.Eval.build(Eval.java:88)
	at io.crate.planner.operators.CorrelatedJoin.build(CorrelatedJoin.java:102)
	at io.crate.planner.operators.Eval.build(Eval.java:83)
	at io.crate.planner.operators.RootRelationBoundary.build(RootRelationBoundary.java:60)
	at io.crate.planner.operators.LogicalPlanner.getNodeOperationTree(LogicalPlanner.java:621)
	at io.crate.planner.operators.LogicalPlanner.doExecute(LogicalPlanner.java:600)
	at io.crate.planner.operators.LogicalPlanner.execute(LogicalPlanner.java:570)
	at io.crate.planner.operators.LogicalPlan.executeOrFail(LogicalPlan.java:216)
	at io.crate.planner.Plan.execute(Plan.java:73)
	at io.crate.session.Session.singleExec(Session.java:797)
	at io.crate.session.Session.execute(Session.java:561)
	at io.crate.rest.action.SqlHttpHandler.executeSimpleRequest(SqlHttpHandler.java:269)
	at io.crate.rest.action.SqlHttpHandler.handleSQLRequest(SqlHttpHandler.java:216)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:124)
	... 50 more
Caused by: java.lang.IllegalArgumentException: Couldn't find (SELECT true FROM (s)) in SourceSymbols{inputs={relation_id=INPUT(1), relation_id=INPUT(0)}, nonDeterministicFunctions={}}
	... 69 more
```

### Expected Result

Successful execution.

**Comment by BaurzhanSakhariev** [comment]
Hi @hammerhead, thanks for the report.

Workaround: 

`SET optimizer_reorder_hash_join = false;`

If you you `NestedLoopJoin` in the plan, shown in the error, then you could use
`SET optimizer_reorder_nested_loop_join = false` to mitigate this.


Issue is related to reorder optimization creating an updated plan with swapped outputs and creating an invalid Eval operator. 
And reorder  optimization, in turn, depends on stats.

**Comment by BaurzhanSakhariev** [comment]
A fix will be available in the next 5.9.4 hotfix release
